onnxruntime/core/providers/cuda/tensor/transpose_impl.cu

```diff
@@ -7,6 +7,167 @@
+  __shared__ T tile[TILE_DIM * (TILE_DIM+1)];
+
+  int x = blockIdx.x * TILE_DIM + threadIdx.x;
+  int y = blockIdx.y * TILE_DIM + threadIdx.y;
+
+  for (int j = 0; j < TILE_DIM; j += BLOCK_ROWS) {
+    tile[(threadIdx.y+j) * TILE_DIM + threadIdx.x] = input_data[blockIdx.z * input_strides[0] + (y+j)*input_shape[2] + x];
+  }
+  __syncthreads();
+
+  x = blockIdx.y * TILE_DIM + threadIdx.x;
+  y = blockIdx.x * TILE_DIM + threadIdx.y;
+
+  for (int j = 0; j < TILE_DIM; j += BLOCK_ROWS) {
+    output_data[blockIdx.z * input_strides[0] + (y+j)*input_shape[1] + x] = tile[threadIdx.x * TILE_DIM + threadIdx.y + j];
+  }
+}
+
+bool canDoTranspose3D(int32_t rank,
+                      const std::vector<int64_t>& input_dims,
+                      const std::vector<size_t>& permutations) {
+  if (rank == 3 &&
+      // permutation is done in the last two dimensions.
+      permutations[rank-2] == (rank-1) && permutations[rank-1] == (rank-2) &&
+      // the last two dimensions are aligned with TILE_DIM.
+      input_dims[rank-2] % TILE_DIM == 0 && input_dims[rank-1] % TILE_DIM == 0) {
+    return true;
+  }
+  return false;
+}
+
+Status Transpose3DImpl(size_t element_size,
+                       const TArray<int64_t>& input_shape, const TArray<int64_t>& input_strides,
+                       const void* input_data, void* output_data, int64_t N) {
+  dim3 block_size(TILE_DIM, BLOCK_ROWS);
+  dim3 grid_size(input_shape[2]/TILE_DIM, input_shape[1]/TILE_DIM, input_shape[0]);
+
+  switch (element_size) {
+    case sizeof(int8_t):
+      _Transpose3DKernel<int8_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int8_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int8_t>::MappedType*>(output_data));
+      break;
+    case sizeof(int16_t):
+      _Transpose3DKernel<int16_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int16_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int16_t>::MappedType*>(output_data));
+      break;
+    case sizeof(int32_t):
+      _Transpose3DKernel<int32_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int32_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int32_t>::MappedType*>(output_data));
+      break;
+    case sizeof(int64_t):
+      _Transpose3DKernel<int64_t><<<grid_size, block_size, 0>>>(
+          input_shape, input_strides,
+          reinterpret_cast<const ToCudaType<int64_t>::MappedType*>(input_data),
+          reinterpret_cast<ToCudaType<int64_t>::MappedType*>(output_data));
+      break;
+    default:
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Type not supported for transpose on CUDA. Element size was ",
+                              element_size);
+  }
+
+  return Status::OK();
+}
+
+template <typename T, int element_size>
```

Comment:
>typename T [](start = 10, length = 10)

how about just copying 4*sizeof(int) bytes from void*'s? what's the need for the intermediate casts to T and int4?

Reply:
int4 is a vector data type provided by CUDA.

Reply:
maybe i'm missing something but i still don't understand why using int4 is better than copying the data directly.

Reply:
with int4, 16 bytes can be copied in one shot.